# 4.7 Deriving New Variables

## Contents:

#### 1. If-Statements with User-Defined Functions

#### 2. If-Statements with the loc() Function

#### 3. If-Statements with For-Loops

#### 4. Creating a New Column

#### 5. Exercise 4.7 Taskwork

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Import dataframe

path = r'C:\Users\jenci\Documents\Data Analytics\Python Fundamentals for Data Analysts\04-2024 Instacart Basket Analysis'
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [3]:
# Check output

df_ords_prods_merge

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32434207   1320836   202557            17                   2   
32434208     31526   202557            18                   5   
32434209    758936   203436             1                   2   
32434210   2745165   203436             2                   3   
32434211   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     7.0       196.0   
1                         7                    15.0       196.0   
2                        12                    21.0       196.0   
3                         7                    29.0       196.0   
4                        15                    28.0       196.0   
...                     ...                     ...         ...   
32434207                 15                     1.0     43553.0   
32434208                 11                     3.0     43553.0   
32434209                  7                     7.0     42338.0   
32434210                  5                    15.0     42338.0   
32434211                 14                     7.0     28818.0   

          add_to_cart_order  reordered                   product_name  \
0                       1.0        0.0                           Soda   
1                       1.0        1.0                           Soda   
2                       1.0        1.0                           Soda   
3                       1.0        1.0                           Soda   
4                       1.0        1.0                           Soda   
...                     ...        ...                            ...   
32434207                2.0        1.0            Orange Energy Shots   
32434208                2.0        1.0            Orange Energy Shots   
32434209                4.0        0.0          Zucchini Chips, Pesto   
32434210               16.0        1.0          Zucchini Chips, Pesto   
32434211                8.0        0.0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge  
0               77              7     9.0   both  
1               77              7     9.0   both  
2               77              7     9.0   both  
3               77              7     9.0   both  
4               77              7     9.0   both  
...            ...            ...     ...    ...  
32434207        64              7     3.7   both  
32434208        64              7     3.7   both  
32434209        50             19     6.9   both  
32434210        50             19     6.9   both  
32434211       130             14    10.3   both  

[32434212 rows x 14 columns]

In [4]:
# Check shape to ensure consitency 

df_ords_prods_merge.shape

(32434212, 14)

### 1. If-Statements with User-Defined Functions

In [5]:
# Create a subset of the df

df = df_ords_prods_merge[:1000000]

In [6]:
# Define function for price labels

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High-range product'
  else: return 'Not enough data'

In [7]:
# Apply price range function on subset dataframe

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\jenci\AppData\Local\Temp\ipykernel_10560\1548173401.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [8]:
# Check values in new column

df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [9]:
# Find maximum label price

df['prices'].max()

14.8

### 2. If-Statements with the loc() Function

In [10]:
# Use the loc() function on subset

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\jenci\AppData\Local\Temp\ipykernel_10560\1095589585.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [11]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [12]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [13]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [14]:
# Use loc() function on data set - loc() runs much faster than the user-defined function b/c it applies the conditional filters before searching through the dataframe, whereas the user-defined function determiens where to set filters after

df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [15]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [16]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21890146
Low-range product     10126384
High-range product      417682
Name: count, dtype: int64

### 3. If-Statements with For-Loops

In [18]:
# Find frequency of orders_day_of week

df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6209632
1    5665830
6    4500246
2    4217766
5    4209449
3    3844096
4    3787193
Name: count, dtype: int64

In [19]:
# Use for-loop function to a list for assigning day-busyness when an order was placed

result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [20]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

#### 4. Creating a New Column

In [21]:
# Create new column in df that contains result

df_ords_prods_merge['busiest_day'] = result

In [22]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22437387
Busiest day        6209632
Least busy         3787193
Name: count, dtype: int64

# 5. Exercise 4.7 Task Work

### Step 2

In [23]:
# Create loop-function for containing new criteria 'busiest days' and ' slowest days'

result_2 = []  # Initialize result_2 as an empty list

for value in df_ords_prods_merge["orders_day_of_week"]:
    if value == 0 or value == 1:
        result_2.append("Busiest days")
    elif value == 4 or value == 3:
        result_2.append("Slowest days")
    else:
        result_2.append("Regularly busy")

In [24]:
# Check list output
result_2

['Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest 

In [25]:
# Create a new column in the df

df_ords_prods_merge['busiest_days'] = result_2

### Step 3 

In [26]:
# Check count
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12927461
Busiest days      11875462
Slowest days       7631289
Name: count, dtype: int64

The sum of all groupings totals to the original number of observations in the dataset (32434212). We can check that the script did exactly what we wanetd by taking the sum of the original order_days_of week count for the 'busiest days', 0 and 1 (6209632 + 5665830) and find that they total to a count of 11875462. Similar can be said about the 'slowest days', 3 and 4.

### Step 4

In [27]:
# Find frequency of order_hour_of_day

df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2764390
11    2738585
14    2691563
15    2664522
13    2663272
12    2620800
16    2537469
9     2456661
17    2089452
8     1719952
18    1637922
19    1259382
20     977017
7      891928
21     796362
22     634737
23     402612
6      290770
0      218942
1      115786
5       88057
2       69431
4       53283
3       51317
Name: count, dtype: int64

In [28]:
# Use for-loop function to create list that categorizes order_hour_of_day into 'most', 'fewest', and 'average orders

result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [29]:
# Create new column containing list of new categories

df_ords_prods_merge['busiest_period_of_day'] = result_3

In [30]:
# Check output
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

### Step 5

In [31]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21137262
Average orders    10006752
Fewest orders      1290198
Name: count, dtype: int64

### Step 7

In [32]:
# Export df
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_2.pkl'))